From 03e7fab7b1cda2b7d674d7706dff89406943508c Mon Sep 17 00:00:00 2001
From: Guoying Zhang <Guoying.Zhang@csr.com>
Date: Thu, 2 Apr 2015 13:46:28 +0800
Subject: [PATCH] modify for udevd to make user can use mount dev

Signed-off-by: Guoying Zhang <Guoying.Zhang@csr.com>
---
 units/systemd-udevd.service.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/units/systemd-udevd.service.in b/units/systemd-udevd.service.in
index f6acd6f..5f8e4c8 100644
--- a/units/systemd-udevd.service.in
+++ b/units/systemd-udevd.service.in
@@ -21,4 +21,4 @@ Sockets=systemd-udevd-control.socket systemd-udevd-kernel.socket
 Restart=always
 RestartSec=0
 ExecStart=@rootlibexecdir@/systemd-udevd
-MountFlags=slave
+MountFlags=shared
-- 
1.9.1

