From b153a05a1ab0ac05688b222d6e2310b8207c8a2b Mon Sep 17 00:00:00 2001
From: Guoying Zhang <Guoying.Zhang@csr.com>
Date: Thu, 26 Mar 2015 12:27:29 +0800
Subject: [PATCH] udev: use other log api to fix build error

Signed-off-by: Guoying Zhang <Guoying.Zhang@csr.com>
---
 src/udev/udev-builtin-kmod.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/udev/udev-builtin-kmod.c b/src/udev/udev-builtin-kmod.c
index 454ee2e..66807b3 100644
--- a/src/udev/udev-builtin-kmod.c
+++ b/src/udev/udev-builtin-kmod.c
@@ -66,7 +66,7 @@ static int load_module(struct udev *udev, const char *alias) {
 _printf_(6,0)
 static void udev_kmod_log(void *data, int priority, const char *file, int line,
                           const char *fn, const char *format, va_list args) {
-        udev_main_log(data, priority, file, line, fn, format, args);
+        log_metav(priority, file, line, fn, format, args);
 }
 
 static int builtin_kmod(struct udev_device *dev, int argc, char *argv[], bool test) {
-- 
1.9.1

